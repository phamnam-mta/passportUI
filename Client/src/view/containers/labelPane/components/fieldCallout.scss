.field-callout-container {
    .textfield-container {
        padding: 10px;
    }
}
