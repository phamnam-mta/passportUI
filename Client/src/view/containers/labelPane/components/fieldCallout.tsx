import * as React from "react";
import { Callout, ICalloutProps, TextField, Stack } from "@fluentui/react";

import "./fieldCallout.scss";

interface IFieldCalloutProps extends ICalloutProps {
    width?: number;
    onCreateField: (value) => void;
    onDismiss: () => void;
    onGetErrorMessage: (value) => string | undefined;
}

export class FieldCallout extends React.PureComponent<IFieldCalloutProps> {
    private onTextFieldKeyDown = (event) => {
        const { onCreateField, onDismiss, onGetErrorMessage } = this.props;
        const hasError = onGetErrorMessage(event.target.value) !== undefined;

        if (event.key === "Enter" && !hasError) {
            onCreateField(event.target.value);
            onDismiss();
        }
    };

    public render() {
        const { width, onDismiss, onGetErrorMessage, ...restProps } = this.props;

        return (
            <Callout
                styles={{ calloutMain: { width } }}
                isBeakVisible={false}
                setInitialFocus
                onDismiss={onDismiss}
                {...restProps}
            >
                <Stack className="field-callout-container" grow={1}>
                    <div className="textfield-container">
                        <TextField
                            onGetErrorMessage={onGetErrorMessage}
                            placeholder="Create new field and hit enter"
                            onKeyDown={this.onTextFieldKeyDown}
                            autoComplete="off"
                        />
                    </div>
                </Stack>
            </Callout>
        );
    }
}

export default FieldCallout;
