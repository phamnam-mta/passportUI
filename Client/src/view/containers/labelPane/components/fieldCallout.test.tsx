import { shallow } from "enzyme";
import * as React from "react";
import { FieldCallout } from "./fieldCallout";

describe("<FieldCallout />", () => {
    let baseProps;
    beforeEach(() => {
        baseProps = {
            width: 200,
            onCreateField: jest.fn(),
            onDismiss: jest.fn(),
            onGetErrorMessage: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            const wrapper = shallow(<FieldCallout {...baseProps} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should call onCreateField and onDismiss, when onKeyDown of TextField is triggered with enter", () => {
            const mockEvent = { key: "Enter", target: { value: "value" } };

            const wrapper = shallow(<FieldCallout {...baseProps} />);
            const textField = wrapper.find("StyledTextFieldBase") as any;
            textField.prop("onKeyDown")(mockEvent);

            expect(baseProps.onCreateField).toBeCalledTimes(1);
            expect(baseProps.onCreateField).toBeCalledWith(mockEvent.target.value);
            expect(baseProps.onDismiss).toBeCalledTimes(1);
        });

        it("should not call onCreateField and onDismiss, when onKeyDown of TextField is triggered with enter with error", () => {
            const props = { ...baseProps, onGetErrorMessage: jest.fn().mockReturnValue("error") };
            const mockEvent = { key: "Enter", target: { value: "value" } };

            const wrapper = shallow(<FieldCallout {...props} />);
            const textField = wrapper.find("StyledTextFieldBase") as any;
            textField.prop("onKeyDown")(mockEvent);

            expect(baseProps.onCreateField).not.toBeCalled();
            expect(baseProps.onDismiss).not.toBeCalled();
        });

        it("should call onGetErrorMessage, when onGetErrorMessage of TextField is triggered", () => {
            const mockValue = "value";

            const wrapper = shallow(<FieldCallout {...baseProps} />);
            const textField = wrapper.find("StyledTextFieldBase") as any;
            textField.prop("onGetErrorMessage")(mockValue);

            expect(baseProps.onGetErrorMessage).toBeCalledTimes(1);
            expect(baseProps.onGetErrorMessage).toBeCalledWith(mockValue);
        });
    });
});
