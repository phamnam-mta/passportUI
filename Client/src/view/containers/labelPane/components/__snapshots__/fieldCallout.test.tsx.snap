// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldCallout /> Rendering should render correctly 1`] = `
<Callout
  isBeakVisible={false}
  onCreateField={[MockFunction]}
  onDismiss={[MockFunction]}
  setInitialFocus={true}
  styles={
    Object {
      "calloutMain": Object {
        "width": 200,
      },
    }
  }
>
  <Stack
    className="field-callout-container"
    grow={1}
  >
    <div
      className="textfield-container"
    >
      <StyledTextFieldBase
        autoComplete="off"
        onGetErrorMessage={[MockFunction]}
        onKeyDown={[Function]}
        placeholder="Create new field and hit enter"
      />
    </div>
  </Stack>
</Callout>
`;
